flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/AbstractDialect.java

```diff
@@ -18,80 +18,252 @@
 
 package org.apache.flink.connector.jdbc.dialect;
 
-import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.annotation.PublicEvolving;
 import org.apache.flink.table.api.ValidationException;
 import org.apache.flink.table.types.DataType;
 import org.apache.flink.table.types.logical.DecimalType;
 import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.RowType;
 import org.apache.flink.table.types.logical.TimestampType;
 import org.apache.flink.table.types.logical.VarBinaryType;
+import org.apache.flink.util.Preconditions;
 
-import java.util.List;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
 
-abstract class AbstractDialect implements JdbcDialect {
+import static java.lang.String.format;
+
+/**
+ * Base class for {@link JdbcDialect JdbcDialects} that implements basic data type validation and
+ * the construction of basic {@code INSERT}, {@code UPDATE}, {@code DELETE}, and {@code SELECT}
+ * statements.
+ *
+ * <p>Implementors should be careful to check the default SQL statements are performant for their
+ * specific dialect and override them if necessary.
+ */
+@PublicEvolving
+public abstract class AbstractDialect implements JdbcDialect {
```

Comment:
While I like this better than it was originally now, I thought about this class a bit more (sorry). With the current changes, this class basically does two things now: 

* It provides default implementations for most methods. Since these are generic and sane, just not necessarily optimized, they make a good fit for a base implementation IMO.
* It performs some level of validation, but really it just implements `validate` and require to implement various other methods instead, essentially changing the way the interface works.

Is there any actual benefit we get out of the latter point? It seems to be that we could just ditch `validate` and the abstract methods of this class and move them into the specific implementations instead without really creating any duplication or additional effort. That would turn this class into a base class which provides generic default implementations. WDYT?

Reply:
> having a generic validation method gives people a level of freedom.

I feel like there's a misunderstanding. I'm not suggesting to remove `validate` as a functionality, but just to remove it from the abstract implementation (and keep it in the interface). In order for any concrete implementation to validate anything, they need to override other methods, but at that point they could just implement `validate` instead.

If we could have a default implementation for `validate` that makes sense across databases, that'd be different. But right now that isn't the case – for the default `validate` implemented here to do _anything_, other methods need to be implemented. What's the benefit of that?

Reply:
I understand you are not saying get rid of `validate`.

I am trying to come up with a trade-off between expressibility and usability. I'd also like to basically ignore the fact we have some dialects implemented in tree and think about someone building a custom dialect.  

The average dialect implementor is not going to have a lot of Flink experience. This is probably their first Flink application and they are someone who has a product requirement to get something working with their DB. What we want is to ensure they are successful so they continue using Flink and take it to production. 

Having this default implementation significantly lowers the barrier to entry. The user lists the data types supported by their DB and gets something sane out of the box. We are on the same page that it does not work generically for all databases, I think it works well enough that the improved user experience outweighs the costs.

Reply:
I actually forgot about the validation of supported types when I wrote that, and I agree with you that that is worth keeping since it applies universally to all implementations. My comment was more directed at the precision validation and the separate API methods introduced for that where I think they don't enough benefit to warrant their own APIs.

Reply:
If you have a better structure I'm open to changing it but LogicalTypeRoot is an enum so we can't put the range in the list. I also don't think we can skip range validation. I have updated the javadoc and exception messages to clearly direct the implementor on what to do to ensure successful validation.

Reply:
I don't have a better suggestion of representing this idea if we insist on keeping it in, but also e.g. `Range` feels out of place for a class like this. I think we're missing a broader concept in the type system here to express "type ranges", and we're working around it by introducing localized, highly specific – but public – APIs for it. But before we start beating a dead horse, it's also public-evolving, implementing a custom JDBC dialect will be a fairly rare occurrence and if we ever have something better to switch this to, the migration path would likely be trivial, so my concern with it is limited; so let's consider this point closed.